evm/src/Hololocker.sol

```diff
@@ -0,0 +1,93 @@
+    event Unlock(address indexed token, address indexed owner, uint256 tokenId, address operator);
+    event Withdraw(address indexed token, address indexed owner, uint256 tokenId, address operator);
+    event LockTimeUpdate(uint256 newValue);
+
+    error InvalidLockTime();
+    error NotUnlockedYet();
+    error TokenNotLocked();
+    error Unauthorized();
+    error UnlockAlreadyRequested();
+
+    // Permits modifications only by the owner of the specified identity.
+    modifier authorized(address token, uint256 tokenId) {
+        _authorized(token, tokenId);
+        _;
+    }
+
+    function lock(address token, uint256 tokenId) external {
+        IERC721(token).transferFrom(msg.sender, address(this), tokenId);
+        nftLockInfo[token][tokenId].owner = msg.sender;
+        nftLockInfo[token][tokenId].operator = msg.sender;
+        emit Lock(token, msg.sender, tokenId, msg.sender);
+    }
+
+    /// @dev Since only authorized user can use this function, it cannot be used without locking NFT beforehand
+    function requestUnlock(address token, uint256 tokenId) external authorized(token, tokenId) {
+        if (nftLockInfo[token][tokenId].unlockTime != 0) {
+            revert UnlockAlreadyRequested();
+        }
+        nftLockInfo[token][tokenId].unlockTime = block.number + lockTime;
+        emit Unlock(token, nftLockInfo[token][tokenId].owner, tokenId, nftLockInfo[token][tokenId].operator);
+    }
+
+    function withdraw(address token, uint256 tokenId) external authorized(token, tokenId) {
+        if (nftLockInfo[token][tokenId].unlockTime == 0 || block.number < nftLockInfo[token][tokenId].unlockTime) {
+            revert NotUnlockedYet();
+        }
+        IERC721(token).transferFrom(address(this), nftLockInfo[token][tokenId].owner, tokenId);
+        emit Withdraw(token, nftLockInfo[token][tokenId].owner, tokenId, nftLockInfo[token][tokenId].operator);
+        delete nftLockInfo[token][tokenId];
+    }
+
+    /// @dev Handles initiating a lock upon direct NFT safeTransferFrom function call
+    function onERC721Received(address operator, address from, uint256 tokenId, bytes calldata)
+        external
+        returns (bytes4)
+    {
+        address token = msg.sender;
+        nftLockInfo[token][tokenId].owner = from;
+        nftLockInfo[token][tokenId].operator = operator;
+        emit Lock(token, from, tokenId, operator);
+        return IERC721Receiver.onERC721Received.selector;
+    }
+
+    function setLockTime(uint256 newLockTime) external onlyOwner {
+        if (newLockTime < MINIMUM_LOCK_TIME) {
+            revert InvalidLockTime();
+        }
+        lockTime = newLockTime;
+        emit LockTimeUpdate(newLockTime);
+    }
+
+    function _authorized(address token, uint256 tokenId) internal view {
+        if (msg.sender != nftLockInfo[token][tokenId].owner && msg.sender != nftLockInfo[token][tokenId].operator) {
```

Comment:
There are two ways to approve [ERC721](https://docs.openzeppelin.com/contracts/3.x/api/token/erc721):
1. `approve(address to, uint256 tokenId)` - Gives permission to to to transfer tokenId token to another account. The approval is cleared when the token is transferred. Only a single account can be approved at a time. The caller must own the token or be an approved operator. This is tracked as `mapping(uint256 tokenId => address) _tokenApprovals`
2. `setApprovalForAll(address operator, bool _approved)` - Approve or remove operator as an operator for the caller. Operators can call transferFrom or safeTransferFrom for any token owned by the caller. The operator cannot be the caller. This is tracked as `mapping(address owner => mapping(address operator => bool)) _operatorApprovals`, so the approval is no longer valid when user transfers the token to another account (since the approval is explicitly from the previous owner).  

So yes, it is possible to have multiple operators, which our system here does not really support tracking. While we could technically query for the single approved address of `_tokenApprovals` at the lock point and add it to authorized users (since that can be different than the address initializing the lock), there is no way to get a list of operators for a certain tokenId, since `_operatorApprovals` is a mapping of a mapping.

So we could support the single approved address to be tracked as an authorized user, but to be honest I don't see much value in that, since it is most common for users to set certain services as the operators rather than the single approved address. Also we would have to call an external contract's function (`getApproved(uint256 tokenId)`), which may pose some security risks we would have to tackle.

Reply:
Okay, if we want to take this position, the most important thing is to document it clearly